### ISSUE TITLE
Unresolved conflict between rules as applied to StyleCopAnalyzers

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#469]
Currently StyleCop.Analyzers has several fields named `_supportedDiagnostics`. Warning SA1309 (FieldNamesMustNotBeginWithUnderscore) is reported for these fields. If the code fix for SA1309 is applied, each of these fields is renamed to `supportedDiagnostics`, which does not cause any errors. However, since each of these fields is marked `static readonly`, the new name results in warning SA1311 (StaticReadonlyFieldsMustBeginWithUpperCaseLetter) getting reported. If the code fix for this rule is applied, the new name of the field is `SupportedDiagnostics`. This name conflicts with the existing property name `SupportedDiagnostics`.

This issue is currently open for discussions regarding potential solutions for addressing this situation now as well as a policy for handling similar naming conflicts in the future.

**Comment by sharwell** [comment]
@pdelvo @macpak Please feel free to add descriptions of possible solutions. We'll discuss them all and decide which one to go with. :+1:

**Comment by sharwell** [comment]
## Possible solution: Suppress warnings

In situations like this, SA1311 could be suppressed through the use of a `#pragma` directive or the `SuppressMessage` attribute.

Advantages:

* This allows the field name to clearly represent its relation to the property with the same name (with the exception of the case of the first letter)

Disadvantages:

* It means we are not following StyleCop rules within a project that *implements StyleCop rules*
* It makes the code ugly
* It means there is more work involved with developing the code, and maintenance costs may rise since otherwise simple operations like renaming a variable involve examining the warning suppressions associated with that variable

**Comment by sharwell** [comment]
### Possible solution: Getter-only auto-property with inline initializer

#### Example:

Before:

```csharp
private static readonly ImmutableArray<DiagnosticDescriptor> supportedDiagnostics =
    ImmutableArray.Create(Descriptor);

public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
{
    get
    {
        return supportedDiagnostics;
    }
}
```

After:

```csharp
public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
{
    get;
} = ImmutableArray.Create(Descriptor);
```

#### Advantages:

* Eliminates the problem and provides even tighter association of the underlying value with the property that exposes it
* Does not result in any additional runtime overhead, unlike some other options like eliminating the cached variable or expression-bodied members

#### Disadvantages:

* Might not increase readability
* Does not provide a general solution or guidance for users who face this problem in the future
* Relies on a C# 6 feature, so isn't even usable as a solution for users working with earlier versions of C#

**Comment by powerdude** [comment]
It would seem like multiple runs would need to be done when doing the code fix.  Somehow SA1309 would need to know it could trigger SA1311.  After SA1309 is done, it could run the process again and see if SA1311 was triggered on the same line.  In this case, it could either reverse the last fix and offer to redo it with a suffix on the field, or just quit and let the user know of the "future" conflict.

**Comment by sharwell** [comment]
### Possible solution: Update SA1311 to ignore these cases

SA1311 could be updated to identify cases where a field `someName` and property `SomeName` which exposes the field exist in the same class. The "exposes the field" determination could either be an assumption (assume that if the field and property exist, that a reasonable user would have implemented the property specifically to expose the field), or it could be based on semantic analysis of the property getter to see if it actually ends with `return this.someName;` (or `return someName;`).

#### Advantages

* Addresses concerns for users with C# 6 as well as with previous versions without requiring code changes or suppressing warnings

#### Disadvantages

* Likely conflicts with original StyleCop behavior
* Potentially defeats one of the consistency goals of StyleCop

**Comment by sharwell** [comment]
### Possible solution: Change to instance field

If the `private static readonly` fields were changed to `private readonly` fields, the naming rules would enforce the use of a lowercase letter instead of an uppercase letter.

#### Advantages

* Easy, low-maintenance change
* Arguably makes sense since the exposed property is also an instance member

#### Disadvantages

* Requires additional allocations if the enclosing type is instantiated multiple times
* Does not provide guidance if the exposing property is `static`

**Comment by sharwell** [comment]
This is a major issue, responsible for hundreds of warnings in the current build (all of them except one in fact).

**Comment by pdelvo** [comment]
I like the solution to update SA1311 to ignore these cases. I would not try to use static analysis to check if the field is actually a backing field because you might have cases where you have multiple return statements. E.g. Return some default value if the field is not initialized. Or return a copy of the field or maybe an immutable version.

**Comment by sharwell** [comment]
I'm not particularly opposed to that, but before deviating from StyleCop on this one (in this manner) I'd like to get a wider consensus.

I was thinking about this again last night, and decided the best name for these should we decide to follow the current rule would be `SomeNameValue`. The code fix could even be updated to detect the existence of a conflicting member name (unlike method names this is extremely easy for fields) and change the proposal to from "Rename to 'SomeName'" to "Rename to 'SomeNameValue'" in the particular case where the field is a backing field for a property (or property-like method).

**Comment by pdelvo** [comment]
I don't think so. I like that idea. But what should happen if there is already a XValue field/property? What if the Property is valled Value? I don't think renaming to eventually FooValueValueValue is a good idea. Maybe we should not offer a diagnostic if this is the case?

**Comment by pdelvo** [comment]
If this is what we are going to do I could implement the code fix and fix the diagnostics in the project itself.

**Comment by sharwell** [comment]
@pdelvo We won't be able to provide a solution in *all* cases, but it does allow us to address one particularly common case. Eventually the general solution for #436 and #437 (whatever that may be) will come into play.

I'll assign this to you for updating the existing code fix to add the `Value` suffix. :+1: 

Don't worry about actually renaming all the fields. I'll do that when I enable the next analyzer release in this project itself. :smile:

**Comment by pdelvo** [comment]
Im going to add Value as often as needed because we send the user into rename mode anyway. I ended up using semantics for this because it was not as trivial as I thought at the beginning. A different part of a partial class, or a base type could define a member of that name.

**Comment by sharwell** [comment]
:bulb: Keep it simple to start with, because there's a pretty high likelihood I'll at least want to *discuss* the possibility of changes to the algorithm prior to a merge (even if we go with the original implementation at the end).

**Comment by macpak** [comment]
I was away for some time, but I see that you agreed on the solution :) Good job :+1:

**Comment by pdelvo** [comment]
I think this issue can be closed
